Introduce Controller and MPC classes to ensure modularity and avoid global config files.

Giovadess opened the pull request:
Code working fine in simulation. 
TO DO:

- [x]  check how to instatiates the controller class and what/how pass parameters to init
- [x] add parametrization for different type of controllers?
- [x]  add docstrings

Danfoa commented on quadruped_pympc/controllers/Controllers.py:
```diff
@@ -0,0 +1,543 @@
+        tau=self.tau
+        return tau    
+
+    def reset(self,env):
```
The `QuadrupedEnv` instance should never be passed to this class. Since if we use this class to control the real robot, we should not depend nor use the env instance. 

What each of these functions should have as signatures is the expected observations from the env. I.e, the feet positions, current contact etc

Giovadess replied:
Remove the use of env inside the code and set the correct signature for each controller function input

Danfoa commented on quadruped_pympc/controllers/Controllers.py:
```diff
@@ -0,0 +1,543 @@
+
+    
+
+    def initialize(self):
```
This code should be in the class initializer. Why define a "initialize" function?

Danfoa commented on quadruped_pympc/controllers/Controllers.py:
```diff
@@ -0,0 +1,543 @@
+
+        # Periodic gait generator _________________________________________________________________________
+        gait_name = self.gait_name
+        gait_type, duty_factor, step_frequency = self.get_gait_params(gait_name)
```
In the new version, the gait parameters are defined not in a function but rather in the config file. 

Meaning that the function `get_gait_params` should disappear, and rather we should expect to get in the constructor of the class the `gait_params` or rather already the instance of the `PeriodicGaitGenerator`

Danfoa commented on quadruped_pympc/controllers/Controllers.py:
```diff
@@ -0,0 +1,543 @@
+####
+    
+#------------
+class Controller:
```
I would use the name `SRBController` or `SingleRigidBodyController`, to clearly state to the user what type of control its executed.

Danfoa commented on quadruped_pympc/controllers/Controllers.py:
```diff
@@ -0,0 +1,543 @@
+class Controller:
+    #constructor
+    def __init__(self,
+                 env,
```
The environment should not be present in this class for the sake of modularity

Danfoa commented on quadruped_pympc/controllers/Controllers.py:
```diff
@@ -0,0 +1,543 @@
+                 control_parametrization=None,
+                 sigma_cem_mppi=None,
+                 num_parallel_computations=None,
+                 num_sampling_iterations=None,
```
As there are parameters that are specific to the type of MPC used under the hood, its best if the "MPC" specific parameters are not literally defined in the signature of the constructor, but rather are handled dynamically, something like:

```python
class SRBController:
   def __init__(
         arg_1,  # Arguments that all MPCs use. E.g., optimize_step_freq
         arg_2,
         ...
         mpc_type: str = 'nominal' # 'nominal' / 'input_rates'
         mpc_params: dict[str, Any] = None
      [...] 
      self.initialize_mpc(mpc_type, mpc_params)
   def set_controller(mpc_type, mpc_params):
       
       if mpc_type.lower() =='nominal':
           NominalClass(**mpc_params)  # This ensures each MPC class checks the appropriate parameters are passed, and if any irrelevant parameters are passed the class will through an unexpected argument Exception
   
```

Giovadess replied:
Ok  I will reformatt this. The init class of the controller will take as input the args and then two dictionaries. One with the simulation parameters and the other with the mpc paramters from the mpc.
It llooks something like this:
`    def __init__(self,
                 hip_height=None,
                 legs_order=None,        
                 inertia=None,
                 gait_params=None,
                 lift_off_positions=None,
                 simulation_parameters:dict=None,
                 mpc_parameters:dict=None,
                 ):`
Now each the mpc_prameter dict will be used in the set_contoller function and will call the specific parameters for each MPC. In order to do this I need to modify the classes for the gradient and sampling bassed controller. This I will do between wednesday and tomorrow. Other then that I fixed all the other issue I willp ush a new version so that you guys can check!

Danfoa commented on quadruped_pympc/controllers/Controllers.py:
```diff
@@ -0,0 +1,543 @@
+import os
+import time
+
+# TODO: Ugly hack so people dont have to run the python command specifying the working directory.
```
I think this hack is not needed anymore

Giovadess replied:
Remove this  and other unused dependencies from the code

Danfoa commented on quadruped_pympc/controllers/Controllers.py:
```diff
@@ -0,0 +1,543 @@
+        # TODO: I would suggest to create a DataClass for "BaseConfig" used in the PotatoModel controllers.
+        self.ref_state = {}
+        # Starting contact sequence
+        self.previous_contact = np.array([1, 1, 1, 1])
```
@giulioturrisi 

why these two instances?

giulioturrisi replied:
which one? previous_contact_sequence? to keep track of the switch between stance and swing.

BTW, can we use all lower caps for the name of the file? the class can be capital letters

Giovadess commented:
After discussing with @giulioturrisi  this morning another thin TO DO is to eliminated the LegAttr Class from inside the controllerClass. I have fixed this I will push in the next few days

Giovadess commented:
I tested the nominal mpc with different trots and it works fine.
I covered most of the comments.
Next step is to divide the dictionaries by controller type and passing them to the different controllerclasses (nominal sampling and so on) that will take care of initializing what needed directly inside the function.
Other thing is to simplify/clean up the config file and dictionaries by removing and keeping only what is really needed.
I expect to do this tomorrow and push the code then we could test it a bit and consider the merge

Giovadess commented:
this new commit contains:
- fixed the problem with the full stance that was not working,
- added custom dictionaries for each controller type and fixed the single controllers they do not depend on config anymore (some fixes needed for inertia),

I need to do some clean up and rerun the simulation almost everything seemed to work next commit should be the last one.
